Generate names automatically when requested

smarterclayton opened the pull request:
This does three things:
- Start moving common object behavior into `pkg/api/rest` (especially around create/update) so that it is not duplicated all over the place.  This will ease the transition for simplifying registry and rest as described  in #1451 
- Introduce a common test harness for REST objects in `pkg/api/rest/resttest` that will reduce the massive cut and paste on create.
- Introduce GenerateName, an optional field on names, as proposed in #148

Addresses #148

smarterclayton commented:
This has breaking implications because builds on Tim's work and tighten's validation.  Should not precede his change.

smarterclayton commented:
Updated the comments with some better description of how this is starting to reorganize rest.

thockin commented on pkg/api/errors/validation.go:
```diff
@@ -84,13 +84,29 @@ type ValidationError struct {
-	s := spew.Sprintf("%s: %s '%+v'", v.Field, v.Type, v.BadValue)
-	if v.Detail != "" {
+	var s string
+	if v.Type == ValidationErrorTypeRequired && isEmpty(v.BadValue) {
```
I have had a long-standing desire to actually just remove the value option to NewFieldRequired() - in EVRY SINGLE case* it is verified to be empty before we pass it in anyway.  Let's just get rid of it.

* pkg/config/config.go passes the error value there - this is just WRONG

smarterclayton replied:
Agreed.


> On Jan 25, 2015, at 11:44 PM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/api/errors/validation.go:
> 
> > @@ -84,13 +84,29 @@ type ValidationError struct {
> >  var _ error = &ValidationError{}
> >  
> >  func (v *ValidationError) Error() string {
> > -	s := spew.Sprintf("%s: %s '%+v'", v.Field, v.Type, v.BadValue)
> > -	if v.Detail != "" {
> > +	var s string
> > +	if v.Type == ValidationErrorTypeRequired && isEmpty(v.BadValue) {
> I have had a long-standing desire to actually just remove the value option to NewFieldRequired() - in EVRY SINGLE case* it is verified to be empty before we pass it in anyway. Let's just get rid of it.
> 
> pkg/config/config.go passes the error value there - this is just WRONG
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented:
OMG, why does git say there are only 4 comments. That will teach me to ignore email.  Catching up...

I agree the suffix shouldn't just be integer digits.

An alternative to encoding incompatible names (dots to dashes, etc.) could be hashing. Labels and annotations could be used to encode more complete, untranslated information.

I think we should drive input filtering from field tags (e.g., `readonly`).

Components need to be able to set object status via the API: #2726.

I'm all for applying defaults and validation as early as possible.

I have a lot of negative experience with fat object layers, and would like to avoid that. I've repeatedly seen them become monolithic honeypots of functionality that couple all code dealing with those objects together, and it gradually becomes impossible to deal with the API objects without the object layer, and impossible . The original pullPolicy defaulting approach had that problem. I like the model of filling out API objects explicitly.

smarterclayton commented:
> On Jan 26, 2015, at 8:26 PM, Brian Grant <notifications@github.com> wrote:
> 
> OMG, why does git say there are only 4 comments. That will teach me to ignore email. Catching up...
> 
> I agree the suffix shouldn't just be integer digits.
> 
> An alternative to encoding incompatible names (dots to dashes, etc.) could be hashing. Labels and annotations could be used to encode more complete, untranslated information.
> 
> I think we should drive input filtering from field tags (e.g., readonly).
> 
> Components need to be able to set object status via the API: #2726.
> 
> I'm all for applying defaults and validation as early as possible.
> 
> I have a lot of negative experience with fat object layers, and would like to avoid that. I've repeatedly seen them become monolithic honeypots of functionality that couple all code dealing with those objects together, and it gradually becomes impossible to deal with the API objects without the object layer, and impossible . The original pullPolicy defaulting approach had that problem. I like the model of filling out API objects explicitly.
> 
I think everyone has that experience, and yet I always see fat object layers.  What specifically about object layer were you referring to?

> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented:
Re. breaking up types.go, see #1744.

bgrant0607 commented:
@smarterclayton Touche. I was reacting to an IRL discussion with @thockin about creating "classes" for the API objects, which sounded related to the comment about constructors above, which, admittedly, I skimmed very quickly.

Will look at the actual PR now. :-)

smarterclayton commented:
Was mosly joking, I am 100% incentivized to reject fat objects (which finalizers, mutable status, and create/update patterns of restful objects can go a long way to reducing).

bgrant0607 commented:
How much of this refactoring is really needed in order to get an initial version of name uniquification in, and to move replication controller to the new mechanism?

smarterclayton commented:
Not much, I'm going to split it all up into a new commit, and do a smaller change.  The create flows right now are terrible, and i will have at a minimum two simultaneous changes.  I don't want to duplicate generate code in multiple places though, and so I may push to do the simple create refactoring first.


> On Jan 27, 2015, at 5:34 PM, Brian Grant <notifications@github.com> wrote:
> 
> How much of this refactoring is really needed in order to get an initial version of name uniquification in, and to move replication controller to the new mechanism?
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton committed:
Add name generation and normalize common create flows

Adds `ObjectMeta.GenerateName`, an optional string field that defines
name generation behavior if a Name is not provided.

Adds `pkg/api/rest`, which defines the default Kubernetes API pattern
for creation (and will cover update as well). Will allow registries
and REST objects to be merged by moving logic on api out of those places.

Add `pkg/api/rest/resttest`, which will be the test suite that verifies
a RESTStorage object follows the Kubernetes API conventions and begin
reducing our duplicated tests.

smarterclayton commented:
I split out a required precursor in #3856, we need to land Tims changes, and then I can move forward.  I don't want to copy and paste generation code, and this stuff is pretty messy right now.


> On Jan 27, 2015, at 5:34 PM, Brian Grant <notifications@github.com> wrote:
> 
> How much of this refactoring is really needed in order to get an initial version of name uniquification in, and to move replication controller to the new mechanism?
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton committed:
Use name generation on pods via replication controllers

The generated name is '<controllerName>-%s', unless controllerName-
would be long enough to cause a validation error.

smarterclayton committed:
Fix integration tests to not depend on setting pod.Status on create

Allow the master to have pod/node cache timeouts controlled via a config
flag for integration tests.

Move integration test to '127.0.0.1' so that it correctly returns a health
check, and enable health check testing on the integration test.

smarterclayton commented:
Rebased, slightly cleaner.  First four commits are part of #3856.  I went with "generateName" as the field, and made this work for pods and replication controllers.  I can split up those two commits as necessary.  Also, this has the placeholder logic to handle cleaning status on create (on those resources we must, like Pod.Status.Host).

smarterclayton committed:
Add name generation to services

smarterclayton committed:
Minions should have common logic with create

TODO: disable / document generate names for cluster scoped resources.

smarterclayton committed:
Turn 409 into 500 Try Again Later when using generateName

If a client says they want the name to be generated, a 409 is
not appropriate (since they didn't specify a name). Instead, we
should return the next most appropriate error, which is a 5xx
error indicating the request failed but the client *should* try
again.  Since there is no 5xx error that exactly fits this purpose,
use 500 with StatusReasonTryAgainLater set.

This commit does not implement client retry on TryAgainLater, but
clients should retry up to a certain number of times.

bgrant0607 commented on cmd/integration/integration.go:
```diff
@@ -186,7 +199,7 @@ func startComponents(manifestURL string) (apiServerURL string) {
 	nodeResources := &api.NodeResources{}
 	nodeController := nodeControllerPkg.NewNodeController(nil, "", machineList, nodeResources, cl)
-	nodeController.Run(10 * time.Second)
+	nodeController.Run(5 * time.Second)
```
Intended to be in this PR?

smarterclayton replied:
Yeah, node controller now needs to run in integration test to mark nodes as healthy.  Before, we were completely ignoring that whole path (disgustingly so).


> On Jan 29, 2015, at 8:44 PM, Brian Grant <notifications@github.com> wrote:
> 
> In cmd/integration/integration.go:
> 
> > @@ -186,7 +199,7 @@ func startComponents(manifestURL string) (apiServerURL string) {
> >  
> >  	nodeResources := &api.NodeResources{}
> >  	nodeController := nodeControllerPkg.NewNodeController(nil, "", machineList, nodeResources, cl)
> > -	nodeController.Run(10 * time.Second)
> > +	nodeController.Run(5 * time.Second)
> Intended to be in this PR?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented on pkg/api/generate.go:
```diff
@@ -0,0 +1,76 @@
+	maxGeneratedNameLength = maxNameLength - randomLength
+)
+
+var randomMax = int(math.Pow(10, randomLength))
```
What is randomMax for? And, isn't it wrong now that more characters are used?

smarterclayton replied:
You're right, it's dead code


> On Jan 29, 2015, at 8:47 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/api/generate.go:
> 
> > +
> > +// simpleNameGenerator generates random names.
> > +type simpleNameGenerator struct{}
> > +
> > +// SimpleNameGenerator is a generator that returns the name plus a random integer between 0 and 999999
> > +// when a name is requested. The string is guaranteed to not exceed the length of a standard Kubernetes
> > +// name (63 characters)
> > +var SimpleNameGenerator NameGenerator = simpleNameGenerator{}
> > +
> > +const (
> > +	maxNameLength          = 63
> > +	randomLength           = 5
> > +	maxGeneratedNameLength = maxNameLength - randomLength
> > +)
> > +
> > +var randomMax = int(math.Pow(10, randomLength))
> What is randomMax for? And, isn't it wrong now that more characters are used?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented on pkg/api/rest/create.go:
```diff
@@ -0,0 +1,100 @@
+		return err
+	}
+
+	return errors.NewTryAgainLater(kind, "POST")
```
This is in hope that some object will be deleted before the client retries?

smarterclayton replied:
Just to leave the door open for us to retry in the future, and give up if we're close to a limit, and eventually control the backoff via a header to clients.  Try again will be useful for repl controllers to watch for.


> On Jan 29, 2015, at 8:56 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/api/rest/create.go:
> 
> > +// to generation being unable to pick a valid name.
> > +func CheckGeneratedNameError(strategy RESTCreateStrategy, err error, obj runtime.Object) error {
> > +	if !errors.IsAlreadyExists(err) {
> > +		return err
> > +	}
> > +
> > +	objectMeta, kind, kerr := objectMetaAndKind(strategy, obj)
> > +	if kerr != nil {
> > +		return kerr
> > +	}
> > +
> > +	if len(objectMeta.GenerateName) == 0 {
> > +		return err
> > +	}
> > +
> > +	return errors.NewTryAgainLater(kind, "POST")
> This is in hope that some object will be deleted before the client retries?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented on pkg/api/rest/types.go:
```diff
@@ -0,0 +1,138 @@
+limitations under the License.
+*/
+
+package rest
```
What's the purpose of this file? A comment would be useful.

smarterclayton replied:
Should be to be the default implementation of the mutation logic for each type.  Once we fold registries into rest we can move these into rest.  They don't have to be here, but logic was small enough so far.


> On Jan 29, 2015, at 8:58 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/api/rest/types.go:
> 
> > +Copyright 2014 Google Inc. All rights reserved.
> > +
> > +Licensed under the Apache License, Version 2.0 (the "License");
> > +you may not use this file except in compliance with the License.
> > +You may obtain a copy of the License at
> > +
> > +    http://www.apache.org/licenses/LICENSE-2.0
> > +
> > +Unless required by applicable law or agreed to in writing, software
> > +distributed under the License is distributed on an "AS IS" BASIS,
> > +WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > +See the License for the specific language governing permissions and
> > +limitations under the License.
> > +*/
> > +
> > +package rest
> What's the purpose of this file? A comment would be useful.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented on pkg/api/v1beta1/types.go:
```diff
@@ -307,6 +307,19 @@ type TypeMeta struct {
+	// either return 201 Created or 500 with Reason TryAgainLater indicating a unique name could not
+	// be found in the time allotted, and the client should retry (optionally after the time indicated
+	// in the Retry-After header).
+	GenerateName string `json:"generateName,omitempty" description:"an optional prefix to use to generate a unique name; has the same validation rules as name; optional and overrides name if specified"`
```
Do we really want this to override name? I think this should behave like regular defaulting: a name is generated only if not provided.

smarterclayton replied:
This was another GET -> POST gut feel.  If I want to copy a bunch of things from namespace 1 to 2 that had generateName set (like a replication controller, or a service), do I also want to generate a name in the new namespace?  The argument for is that you'd be able to GET -> POST in the same namespace to copy something.  The argument against is that any interlinks would be broken by name generation (if you exported multiple things).

Which way would you land?

----- Original Message -----
> > @@ -307,6 +307,19 @@ type TypeMeta struct {
> >  	APIVersion        string    `json:"apiVersion,omitempty"
> >  	description:"version of the schema the object should have"`
> >  	Namespace         string    `json:"namespace,omitempty"
> >  	description:"namespace to which the object belongs; must be a
> >  	DNS_SUBDOMAIN; 'default' by default"`
> >  
> > +	// GenerateName indicates that the name should be made unique by the
> > server prior to persisting
> > +	// it. A non-empty value for the field indicates the name will be made
> > unique (and the name
> > +	// returned to the client will be different than the name passed). The
> > value of this field will
> > +	// be combined with a unique suffix on the server, and takes precedence
> > over the Name field.
> > +	// The provided value must be valid within the rules for Name, and may be
> > truncated by the length
> > +	// of the suffix required to make the value unique on the server.
> > +	//
> > +	// If this field is specified, the server will NOT return a 409 if Name
> > exists - instead, it will
> > +	// either return 201 Created or 500 with Reason TryAgainLater indicating
> > a unique name could not
> > +	// be found in the time allotted, and the client should retry (optionally
> > after the time indicated
> > +	// in the Retry-After header).
> > +	GenerateName string `json:"generateName,omitempty" description:"an
> > optional prefix to use to generate a unique name; has the same validation
> > rules as name; optional and overrides name if specified"`
> 
> Do we really want this to override name? I think this should behave like
> regular defaulting: a name is generated only if not provided.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3789/files#r23820525

bgrant0607 replied:
Continuing discussion from #3908:

I would like to make a hard rule that we don't mutate writable fields the client sends. OTOH, we can default unspecified fields all we want, and should make defaults fully explicit in the metadata and spec.

GET -> patch -> PUT (narrow update) needs to work in a straightforward manner.

Submit/apply/ensure (declarative update), extract (create clean config), and clone need to be based on user intent. The plan of record is to use annotations to record user intent rather than an official second layer of desired state.

Cloning groups of objects would indeed be even harder. However, I don't see that as being a likely scenario for objects with independently uniquified names.

bgrant0607 commented:
LGTM, other than removing randomMax.

Which PRs is this still waiting on, if any?

smarterclayton commented:
None.

----- Original Message -----
> LGTM, other than removing randomMax.
> 
> Which PRs is this still waiting on, if any?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3789#issuecomment-72165003

smarterclayton commented:
GenerateName is now only used if Name is not specified.  maxNameLength is required in order to ensure we can generate names from very long prefixes - I added a TODO that if other apis (outside of core) wish to have longer/shorter/different names they will need to change that value.

There are no other outstanding review comments.

bgrant0607 commented:
Thanks much. LGTM. It looks like randomMax is still there, but I don't want to hold this up for that.

Rerunning travis.

smarterclayton commented:
Hold on, thats not the right code being shown.

bgrant0607 commented:
Also, data race detected in pkg/registry/registrytest/controller.go: https://app.shippable.com/builds/54cd100e9821921100fa6aa4

smarterclayton commented:
I'll do a follow up commit to nuke that.  Looks like I squashed the fix for Tims comment into the wrong commit, so they got layered wrong.


> On Jan 31, 2015, at 1:33 PM, Brian Grant <notifications@github.com> wrote:
> 
> Thanks much. LGTM. It looks like randomMax is still there, but I don't want to hold this up for that.
> 
> Rerunning travis.
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton commented:
Will dig in later and fix.


> On Jan 31, 2015, at 1:36 PM, Brian Grant <notifications@github.com> wrote:
> 
> Also, data race detected in pkg/registry/registrytest/controller.go: https://app.shippable.com/builds/54cd100e9821921100fa6aa4
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton commented:
Should be cleaned up with the previous comments all addressed.  Race detector on 1.4 is more sophisticated apparently.

thockin commented on pkg/api/generate.go:
```diff
@@ -0,0 +1,74 @@
+// simpleNameGenerator generates random names.
+type simpleNameGenerator struct{}
+
+// SimpleNameGenerator is a generator that returns the name plus a random integer between 0 and 999999
```
nit: this still describes an integer

thockin commented on pkg/api/types.go:
```diff
@@ -82,6 +82,19 @@ type ObjectMeta struct {
+	// GenerateName indicates that the name should be made unique by the server prior to persisting
+	// it. A non-empty value for the field indicates the name will be made unique (and the name
+	// returned to the client will be different than the name passed). The value of this field will
+	// be combined with a unique suffix on the server if the Name field has not been provided.
```
"if the Name field has not been provided".  What if a user provides name and generateName?

smarterclayton replied:
I thought it was clarified later - Name takes precedence over GenerateName

thockin commented on pkg/api/types.go:
```diff
@@ -82,6 +82,19 @@ type ObjectMeta struct {
+	// of the suffix required to make the value unique on the server.
+	//
+	// If this field is specified, and Name is not present, the server will NOT return a 409 if the
+	// generated name exists - instead, it will/ either return 201 Created or 500 with Reason
```
typo?  will/

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -30,11 +30,53 @@ import (
+// ValidatePodName can be used to check whether the given pod name is valid.
+// Prefix indicates this name will be used as part of generation, in which case
+// trailing dashes are allowed.
+func ValidatePodName(name string, prefix bool) (bool, string) {
```
I don't find muhc value in these functions, vs directly passing nameIsDNSSubdomain at the call-site.  Is there some evolutionary plan that justifies these?

smarterclayton replied:
They're called by the generation code explicitly when we generate names, so that we can separately validate name changes from validating the whole pod.

smarterclayton replied:
See replication_controller.go line 67

thockin replied:
That justifies pod name - none of the others.  It's minor, it's just noise without a reason.

smarterclayton replied:
Just today I made up a "use case" and used these in OpenShift out of spite - we are doing generation of replication controllers for each deployment so we sanity check the replication controller name validation prior to submitting it to the api, and we started generating services and I was letting the service name validation guide whether we made up a name, or used one that corresponded to the docker image's name we generated an app from.  Having them separate was truly useful.


> On Feb 2, 2015, at 6:50 PM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/api/validation/validation.go:
> 
> > +// name will have a value appended to it.
> > +type ValidateNameFunc func(name string, prefix bool) (bool, string)
> > +
> > +// maskTrailingDash replaces the final character of a string with a subdomain safe
> > +// value if is a dash.
> > +func maskTrailingDash(name string) string {
> > +	if strings.HasSuffix(name, "-") {
> > +		return name[:len(name)-2] + "a"
> > +	}
> > +	return name
> > +}
> > +
> > +// ValidatePodName can be used to check whether the given pod name is valid.
> > +// Prefix indicates this name will be used as part of generation, in which case
> > +// trailing dashes are allowed.
> > +func ValidatePodName(name string, prefix bool) (bool, string) {
> That justifies pod name - none of the others. It's minor, it's just noise without a reason.
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton commented:
Rebased

----- Original Message -----
> LGTM. We can fix more later.
> 
> @smarterclayton Please rebase.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3789#issuecomment-72520940

lavalamp commented:
The new resttest package really needed a doc.go file and some godoc comments for the public methods. It's pretty baffling as-is.

smarterclayton commented:
Everyone is a critic :). I was just tired of copying and pasting the same 30 tests over and over.
